meso_cluster_analysis.py

```
@@ -0,0 +1,420 @@
+    """Combine micro posteriors and reconciliation penalties into a cluster view."""
+
+    posts = _to_float_sequence(micro_posteriors)
+    if not posts:
+        raise ValueError("micro_posteriors cannot be empty")
+
+    if weighting_trace is None:
+        weights = [1.0] * len(posts)
+    else:
+        weights = _to_float_sequence(weighting_trace)
+        if len(weights) != len(posts):
+            raise ValueError("weighting_trace must match micro_posteriors length")
```

Ruff (0.14.2)

Avoid specifying long messages outside the exception class

(TRY003)